sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,373 @@
+        /// </summary>
+        private IKeyEncryptionKeyResolver KeyResolver { get; }
+
+        /// <summary>
+        /// An already locally existing IKeyEncryptionKey, for key-fetching optimizations.
+        /// </summary>
+        private IKeyEncryptionKey LocalKey { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ClientSideDecryptionPolicy"/> class with the specified key and resolver.
+        /// <para />
+        /// If <paramref name="key"/> is specified, that key will attempt to be used in decryption, provided it matches the
+        /// content, before using the <paramref name="keyResolver"/> to find the correct key.
+        /// </summary>
+        /// <param name="keyResolver">
+        /// Key resolver for finding the appropriate <see cref="IKeyEncryptionKey"/> to unwrap the blob's content
+        /// encryption key.
+        /// </param>
+        /// <param name="key">
+        /// The <see cref="IKeyEncryptionKey"/> used by this pipeline's client to wrap encryption keys on uploads, if present.
+        /// Used to skip a key fetch with <paramref name="keyResolver"/> if unnecessary.
+        /// </param>
+        public ClientSideDecryptionPolicy(IKeyEncryptionKeyResolver keyResolver = default, IKeyEncryptionKey key = default)
+        {
+            LocalKey = key;
+            KeyResolver = keyResolver;
+        }
+
+        #region Process
+
+        /// <summary>
+        /// If the request is fetching a specific range from the blob, adjust that range to download the extra bytes
+        /// needed for decryption. On return, pipe the content through a CryptoStream to decrypt the data.
+        /// </summary>
+        /// <param name="message">The <see cref="HttpMessage"/> this policy would be applied to.</param>
+        /// <param name="pipeline">The set of <see cref="HttpPipelinePolicy"/> to execute after current one.</param>
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessImpl(message, pipeline, false).EnsureCompleted();
+        }
+
+        /// <summary>
+        /// If the request is fetching a specific range from the blob, adjust that range to download the extra bytes
+        /// needed for decryption. On return, pipe the content through a CryptoStream to decrypt the data.
+        /// </summary>
+        /// <param name="message">The <see cref="HttpMessage"/> this policy would be applied to.</param>
+        /// <param name="pipeline">The set of <see cref="HttpPipelinePolicy"/> to execute after current one.</param>
+        /// <returns></returns>
+        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            await ProcessImpl(message, pipeline, false).ConfigureAwait(false);
+        }
+
+        private async ValueTask ProcessImpl(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
```

Comment:
Slight nit - this is kind of dumb, but we have an unofficial guideline to not use `Impl` as a suffix because it makes stack traces ugly.  Would you mind switching to `Internal` as a suffix instead?